README.md
# Filecoin Proofs FFI

> C and CGO bindings for Filecoin's Rust libraries

## Building

To build and install libfilcrypto, its header file and pkg-config manifest, run:

```shell
make
```

To optionally authenticate with GitHub for assets download (to increase API limits)
set `GITHUB_TOKEN` to personal access token.

If no precompiled static library is available for your operating system, the
build tooling will attempt to compile a static library from local Rust sources.

### Installation notes

By default, building this will download a pre-built binary of the ffi.  The advantages for downloading it are faster build times, and not requiring a rust toolchain and build environment.

The disadvantage to downloading the pre-built binary is that it will not be optimized for your specific hardware.  This means that if raw performance is of utmost importance to you, it's highly recommended that you build from source.

### Building from Source

To opt out of downloading precompiled assets, set `FFI_BUILD_FROM_SOURCE=1`:

```shell
rm .install-filcrypto \
    ; make clean \
    ; FFI_BUILD_FROM_SOURCE=1 make
```

To allow portable building of the `blst` dependency, set `FFI_USE_BLST_PORTABLE=1`:

```shell
rm .install-filcrypto \
    ; make clean \
    ; FFI_BUILD_FROM_SOURCE=1 FFI_USE_BLST_PORTABLE=1 make
```

By default, a 'gpu' option is used in the proofs library.  There is now an experimental 'gpu2' feature which has improved performance.  To allow building with the 'gpu2' dependency, set `FFI_USE_GPU2=1`:

```shell
rm .install-filcrypto \
    ; make clean \
    ; FFI_BUILD_FROM_SOURCE=1 FFI_USE_BLST=1 FFI_USE_GPU2=1 make
```

## Updating rust-fil-proofs (via rust-filecoin-proofs-api)

If rust-fil-proofs has changed from commit X to Y and you wish to get Y into
the filecoin-ffi project, you need to do a few things:

1. Update the rust-filecoin-proofs-api [Cargo.toml][1] file to point to Y
2. Run `cd rust && cargo update -p "filecoin-proofs-api"` from the root of the filecoin-ffi project
3. After the previous step alters your Cargo.lock file, commit and push

## go get

`go get` needs some additional steps in order to work as expected.

Get the source, add this repo as a submodule to your repo, build it and point to it:

```shell
$ go get github.com/chenjianmei111/filecoin-ffi
$ git submodule add https://github.com/chenjianmei111/filecoin-ffi.git extern/filecoin-ffi
$ make -C extern/filecoin-ffi
$ go mod edit -replace=github.com/chenjianmei111/filecoin-ffi=./extern/filecoin-ffi
```

## Updating CGO Bindings

The CGO bindings are generated using [c-for-go](https://github.com/xlab/c-for-go)
and committed to Git. To generate bindings yourself, install the c-for-go
binary, ensure that it's on your path, and then run `make cgo-gen`. CI builds
will fail if generated CGO diverges from what's checked into Git.

## Updating the Changelog

The `mkreleaselog` script (in the project root) can be used to generate a good
portion of the filecoin-ffi changelog. For historical reasons, the script must
be run from the root of a filecoin-ffi checkout which is in your `$GOPATH`.

Run it like so:

```shell
./mkreleaselog v0.25.0 v0.26.0 > /tmp/v0.26.0.notes.txt
```

## License

MIT or Apache 2.0

[1]: https://github.com/chenjianmei111/rust-filecoin-proofs-api/commit/61fde0e581cc38abc4e13dbe96145c9ad2f1f0f5


CHANGELOG.md
# filecoin-ffi changelog

## 0.30.3

This release adds `FauxRep` to the `ffi` package, and a few other
rust-fil-proofs improvements, which you can read about [here](https://github.com/chenjianmei111/rust-fil-proofs/blob/master/CHANGELOG.md#403---2020-07-01).

### Changelog

- github.com/chenjianmei111/filecoin-ffi:
  - add FauxRep to ffi package (#118) ([chenjianmei111/filecoin-ffi#118](https://github.com/chenjianmei111/filecoin-ffi/pull/118))

### Contributors

| Contributor | Commits | Lines ± | Files Changed |
|-------------|---------|---------|---------------|
| Erin Swenson-Healey | 1 | +370/-143 | 9 |

## 0.30.2

This release fixed a bug in the `ffi.Aggregate` function, which did not properly
handle a NUL byte (representing an error) returned from the `fil_aggregate` Rust
function.

### Changelog

- github.com/chenjianmei111/filecoin-ffi:
  - handle null pointer returned from fil_aggregate (#116) ([chenjianmei111/filecoin-ffi#116](https://github.com/chenjianmei111/filecoin-ffi/pull/116))
  - 0.30.1 changelog (#114) ([chenjianmei111/filecoin-ffi#114](https://github.com/chenjianmei111/filecoin-ffi/pull/114))

### Contributors

| Contributor | Commits | Lines ± | Files Changed |
|-------------|---------|---------|---------------|
| Erin Swenson-Healey | 2 | +45/-1 | 3 |

## 0.30.1

This release wil include Window PoSt speedups (2x measured for 32GiB sectors),
RAM reduction of 56GiB for 32GiB sectors (mmap'd parent cache with windows for
access, rather than all in RAM at once), and some phase2/trusted setup related
updates (for the trusted setup participants).

### Changelog

- github.com/chenjianmei111/filecoin-ffi:
  - update to rust-fil-proofs 4.0.2 (#113) ([chenjianmei111/filecoin-ffi#113](https://github.com/chenjianmei111/filecoin-ffi/pull/113))
  - run the Rust tests before running the Go tests (#112) ([chenjianmei111/filecoin-ffi#112](https://github.com/chenjianmei111/filecoin-ffi/pull/112))
  - Update master dependencies ([chenjianmei111/filecoin-ffi#111](https://github.com/chenjianmei111/filecoin-ffi/pull/111))
  - update changelog for 0.30.0 release ([chenjianmei111/filecoin-ffi#109](https://github.com/chenjianmei111/filecoin-ffi/pull/109))
- github.com/chenjianmei111/specs-actors (v0.6.0 -> v0.6.1)

### Contributors

| Contributor | Commits | Lines ± | Files Changed |
|-------------|---------|---------|---------------|
| nemo | 3 | +1463/-1782 | 17 |
| Alex North | 2 | +139/-116 | 3 |
| Erin Swenson-Healey | 2 | +90/-68 | 7 |
| laser | 1 | +31/-0 | 1 |

## 0.30.0

This release includes an update specs-actors (splits abi.RegisteredProof into
two new types - one for seal and one for PoSt) and an update to rust-fil-proofs
4.0.0, which you can read about [here](https://github.com/chenjianmei111/rust-fil-proofs/blob/master/CHANGELOG.md#400---2020-06-15).

### Changelog

- github.com/chenjianmei111/filecoin-ffi:
  - update to rust-fil-proofs 4.0.0 (#108) ([chenjianmei111/filecoin-ffi#108](https://github.com/chenjianmei111/filecoin-ffi/pull/108))
  - specs-actors v0.6 ([chenjianmei111/filecoin-ffi#107](https://github.com/chenjianmei111/filecoin-ffi/pull/107))
  - changelog for 0.28.0, 0.28.1, and 0.29.0 (#106) ([chenjianmei111/filecoin-ffi#106](https://github.com/chenjianmei111/filecoin-ffi/pull/106))
- github.com/chenjianmei111/specs-actors (v0.5.4-0.20200521014528-0df536f7e461 -> v0.6.0)

### Contributors

| Contributor | Commits | Lines ± | Files Changed |
|-------------|---------|---------|---------------|
| Frrist | 3 | +454/-239 | 22 |
| Whyrusleeping | 2 | +485/-119 | 19 |
| Alex North | 7 | +424/-151 | 15 |
| Łukasz Magiera | 4 | +227/-154 | 13 |
| Alex Cruikshank | 4 | +262/-85 | 10 |
| porcuquine | 1 | +172/-171 | 15 |
| Erin Swenson-Healey | 3 | +153/-30 | 5 |
| ZenGround0 | 3 | +42/-17 | 16 |
| ZX | 2 | +16/-19 | 4 |
| WEI YANG | 1 | +6/-2 | 1 |
| Henri | 1 | +2/-2 | 1 |


## 0.29.0

Big changes here! We moved off of the nightly Rust channel, fixed a nasty file
descriptor-leak, and (most importantly) updated to [v27 parameters and keys](https://github.com/chenjianmei111/rust-fil-proofs/blob/master/CHANGELOG.md#300---2020-06-08).

### Changelog

- github.com/chenjianmei111/filecoin-ffi:
  - fix: update to filecoin-proofs-api v3.0.0 ([chenjianmei111/filecoin-ffi#105](https://github.com/chenjianmei111/filecoin-ffi/pull/105))
  - explicitly close os.File to force release of file descriptor (#97) ([chenjianmei111/filecoin-ffi#97](https://github.com/chenjianmei111/filecoin-ffi/pull/97))
  - fix: use stable 1.43.1 release ([chenjianmei111/filecoin-ffi#102](https://github.com/chenjianmei111/filecoin-ffi/pull/102))

### Contributors

| Contributor | Commits | Lines ± | Files Changed |
|-------------|---------|---------|---------------|
| nemo | 2 | +133/-132 | 4 |
| Erin Swenson-Healey | 1 | +47/-1 | 2 |
| Volker Mische | 1 | +1/-3 | 3 |


## 0.28.1

This release modifies the rust-filecoin-proofs-api dependency, downloading it
from crates.io instead of GitHub. No behavior changes.

### Changelog

- github.com/chenjianmei111/filecoin-ffi:

proofs_test.go
package ffi

import (
	"bytes"
	"crypto/rand"
	"io"
	"io/ioutil"
	"math/big"
	"testing"

	"github.com/chenjianmei111/filecoin-ffi/generated"

	"github.com/stretchr/testify/assert"

	commcid "github.com/chenjianmei111/go-fil-commcid"

	"github.com/chenjianmei111/go-state-types/abi"

	"github.com/stretchr/testify/require"
)

func TestRegisteredSealProofFunctions(t *testing.T) {
	WorkflowRegisteredSealProofFunctions(newTestingTeeHelper(t))
}

func TestRegisteredPoStProofFunctions(t *testing.T) {
	WorkflowRegisteredPoStProofFunctions(newTestingTeeHelper(t))
}

func TestProofsLifecycle(t *testing.T) {
	WorkflowProofsLifecycle(newTestingTeeHelper(t))
}

func TestGetGPUDevicesDoesNotProduceAnError(t *testing.T) {
	WorkflowGetGPUDevicesDoesNotProduceAnError(newTestingTeeHelper(t))
}

func TestGenerateWinningPoStSectorChallenge(t *testing.T) {
	WorkflowGenerateWinningPoStSectorChallenge(newTestingTeeHelper(t))
}

func TestGenerateWinningPoStSectorChallengeEdgeCase(t *testing.T) {
	WorkflowGenerateWinningPoStSectorChallengeEdgeCase(newTestingTeeHelper(t))
}

func TestJsonMarshalSymmetry(t *testing.T) {
	for i := 0; i < 100; i++ {
		xs := make([]publicSectorInfo, 10)
		for j := 0; j < 10; j++ {
			var x publicSectorInfo
			var commR [32]byte
			_, err := io.ReadFull(rand.Reader, commR[:])
			require.NoError(t, err)

			// commR is defined as 32 long above, error can be safely ignored
			x.SealedCID, _ = commcid.ReplicaCommitmentV1ToCID(commR[:])

			n, err := rand.Int(rand.Reader, big.NewInt(500))
			require.NoError(t, err)
			x.SectorNum = abi.SectorNumber(n.Uint64())
			xs[j] = x
		}
		toSerialize := newSortedPublicSectorInfo(xs...)

		serialized, err := toSerialize.MarshalJSON()
		require.NoError(t, err)

		var fromSerialized SortedPublicSectorInfo
		err = fromSerialized.UnmarshalJSON(serialized)
		require.NoError(t, err)

		require.Equal(t, toSerialize, fromSerialized)
	}
}

func TestDoesNotExhaustFileDescriptors(t *testing.T) {
	m := 500         // loops
	n := uint64(508) // quantity of piece bytes

	for i := 0; i < m; i++ {
		// create a temporary file over which we'll compute CommP
		file, err := ioutil.TempFile("", "")
		if err != nil {
			panic(err)
		}

		// create a slice of random bytes (represents our piece)
		b := make([]byte, n)

		// load up our byte slice with random bytes
		if _, err = rand.Read(b); err != nil {
			panic(err)
		}

		// write buffer to temp file
		if _, err := bytes.NewBuffer(b).WriteTo(file); err != nil {
			panic(err)
		}

		// seek to beginning of file
		if _, err := file.Seek(0, 0); err != nil {
			panic(err)
		}

		if _, err = GeneratePieceCID(abi.RegisteredSealProof_StackedDrg2KiBV1, file.Name(), abi.UnpaddedPieceSize(n)); err != nil {
			panic(err)
		}

		if err = file.Close(); err != nil {
			panic(err)
		}
	}
}

func newTestingTeeHelper(t *testing.T) *testingTeeHelper {
	return &testingTeeHelper{t: t}
}

type testingTeeHelper struct {
	t *testing.T
}

func (tth *testingTeeHelper) RequireTrue(value bool, msgAndArgs ...interface{}) {
	require.True(tth.t, value, msgAndArgs)
}

func (tth *testingTeeHelper) RequireNoError(err error, msgAndArgs ...interface{}) {
	require.NoError(tth.t, err, msgAndArgs)
}

func (tth *testingTeeHelper) RequireEqual(expected interface{}, actual interface{}, msgAndArgs ...interface{}) {
	require.Equal(tth.t, expected, actual, msgAndArgs)
}

func (tth *testingTeeHelper) AssertNoError(err error, msgAndArgs ...interface{}) bool {
	return assert.NoError(tth.t, err, msgAndArgs)
}

func (tth *testingTeeHelper) AssertEqual(expected, actual interface{}, msgAndArgs ...interface{}) bool {
	return assert.Equal(tth.t, expected, actual, msgAndArgs)
}

func (tth *testingTeeHelper) AssertTrue(value bool, msgAndArgs ...interface{}) bool {
	return assert.True(tth.t, value, msgAndArgs)
}

func TestProofTypes(t *testing.T) {
	assert.EqualValues(t, generated.FilRegisteredPoStProofStackedDrgWinning2KiBV1, abi.RegisteredPoStProof_StackedDrgWinning2KiBV1)
	assert.EqualValues(t, generated.FilRegisteredPoStProofStackedDrgWinning8MiBV1, abi.RegisteredPoStProof_StackedDrgWinning8MiBV1)
	assert.EqualValues(t, generated.FilRegisteredPoStProofStackedDrgWinning512MiBV1, abi.RegisteredPoStProof_StackedDrgWinning512MiBV1)
	assert.EqualValues(t, generated.FilRegisteredPoStProofStackedDrgWinning32GiBV1, abi.RegisteredPoStProof_StackedDrgWinning32GiBV1)
	assert.EqualValues(t, generated.FilRegisteredPoStProofStackedDrgWinning64GiBV1, abi.RegisteredPoStProof_StackedDrgWinning64GiBV1)
	assert.EqualValues(t, generated.FilRegisteredPoStProofStackedDrgWindow2KiBV1, abi.RegisteredPoStProof_StackedDrgWindow2KiBV1)
	assert.EqualValues(t, generated.FilRegisteredPoStProofStackedDrgWindow8MiBV1, abi.RegisteredPoStProof_StackedDrgWindow8MiBV1)
	assert.EqualValues(t, generated.FilRegisteredPoStProofStackedDrgWindow512MiBV1, abi.RegisteredPoStProof_StackedDrgWindow512MiBV1)
	assert.EqualValues(t, generated.FilRegisteredPoStProofStackedDrgWindow32GiBV1, abi.RegisteredPoStProof_StackedDrgWindow32GiBV1)
	assert.EqualValues(t, generated.FilRegisteredPoStProofStackedDrgWindow64GiBV1, abi.RegisteredPoStProof_StackedDrgWindow64GiBV1)

	assert.EqualValues(t, generated.FilRegisteredSealProofStackedDrg2KiBV1, abi.RegisteredSealProof_StackedDrg2KiBV1)
	assert.EqualValues(t, generated.FilRegisteredSealProofStackedDrg8MiBV1, abi.RegisteredSealProof_StackedDrg8MiBV1)
	assert.EqualValues(t, generated.FilRegisteredSealProofStackedDrg512MiBV1, abi.RegisteredSealProof_StackedDrg512MiBV1)
	assert.EqualValues(t, generated.FilRegisteredSealProofStackedDrg32GiBV1, abi.RegisteredSealProof_StackedDrg32GiBV1)
	assert.EqualValues(t, generated.FilRegisteredSealProofStackedDrg64GiBV1, abi.RegisteredSealProof_StackedDrg64GiBV1)
}


workflows.go
//+build cgo

package ffi

import (
	"bytes"
	"crypto/rand"
	"encoding/binary"
	"fmt"
	"io"
	"io/ioutil"
	"math"
	"math/big"
	"os"
	"path/filepath"

	"github.com/chenjianmei111/go-state-types/abi"
	prf "github.com/chenjianmei111/specs-actors/actors/runtime/proof"
	"github.com/ipfs/go-cid"
)

func WorkflowProofsLifecycle(t TestHelper) {
	minerID := abi.ActorID(42)
	randomness := [32]byte{9, 9, 9}
	sealProofType := abi.RegisteredSealProof_StackedDrg2KiBV1
	winningPostProofType := abi.RegisteredPoStProof_StackedDrgWinning2KiBV1
	sectorNum := abi.SectorNumber(42)

	ticket := abi.SealRandomness{5, 4, 2}

	seed := abi.InteractiveSealRandomness{7, 4, 2}

	// initialize a sector builder
	metadataDir := requireTempDirPath(t, "metadata")
	defer os.RemoveAll(metadataDir)

	sealedSectorsDir := requireTempDirPath(t, "sealed-sectors")
	defer os.RemoveAll(sealedSectorsDir)

	stagedSectorsDir := requireTempDirPath(t, "staged-sectors")
	defer os.RemoveAll(stagedSectorsDir)

	sectorCacheRootDir := requireTempDirPath(t, "sector-cache-root-dir")
	defer os.RemoveAll(sectorCacheRootDir)

	sectorCacheDirPath := requireTempDirPath(t, "sector-cache-dir")
	defer os.RemoveAll(sectorCacheDirPath)

	fauxSectorCacheDirPath := requireTempDirPath(t, "faux-sector-cache-dir")
	defer os.RemoveAll(fauxSectorCacheDirPath)

	stagedSectorFile := requireTempFile(t, bytes.NewReader([]byte{}), 0)
	defer stagedSectorFile.Close()

	sealedSectorFile := requireTempFile(t, bytes.NewReader([]byte{}), 0)
	defer sealedSectorFile.Close()

	fauxSealedSectorFile := requireTempFile(t, bytes.NewReader([]byte{}), 0)
	defer fauxSealedSectorFile.Close()

	unsealOutputFileA := requireTempFile(t, bytes.NewReader([]byte{}), 0)
	defer unsealOutputFileA.Close()

	unsealOutputFileB := requireTempFile(t, bytes.NewReader([]byte{}), 0)
	defer unsealOutputFileB.Close()

	unsealOutputFileC := requireTempFile(t, bytes.NewReader([]byte{}), 0)
	defer unsealOutputFileC.Close()

	unsealOutputFileD := requireTempFile(t, bytes.NewReader([]byte{}), 0)
	defer unsealOutputFileD.Close()

	// some rando bytes
	someBytes := make([]byte, abi.PaddedPieceSize(2048).Unpadded())
	_, err := io.ReadFull(rand.Reader, someBytes)
	t.RequireNoError(err)

	// write first piece
	pieceFileA := requireTempFile(t, bytes.NewReader(someBytes[0:127]), 127)

	pieceCIDA, err := GeneratePieceCIDFromFile(sealProofType, pieceFileA, 127)
	t.RequireNoError(err)

	// seek back to head (generating piece commitment moves offset)
	_, err = pieceFileA.Seek(0, 0)
	t.RequireNoError(err)

	// write the first piece using the alignment-free function
	n, pieceCID, err := WriteWithoutAlignment(sealProofType, pieceFileA, 127, stagedSectorFile)
	t.RequireNoError(err)
	t.AssertEqual(int(n), 127)
	t.AssertTrue(pieceCID.Equals(pieceCIDA))

	// write second piece + alignment
	t.RequireNoError(err)
	pieceFileB := requireTempFile(t, bytes.NewReader(someBytes[0:1016]), 1016)

	pieceCIDB, err := GeneratePieceCIDFromFile(sealProofType, pieceFileB, 1016)
	t.RequireNoError(err)

	// seek back to head
	_, err = pieceFileB.Seek(0, 0)
	t.RequireNoError(err)

	// second piece relies on the alignment-computing version
	left, tot, pieceCID, err := WriteWithAlignment(sealProofType, pieceFileB, 1016, stagedSectorFile, []abi.UnpaddedPieceSize{127})
	t.RequireNoError(err)
	t.AssertEqual(889, int(left))
	t.AssertEqual(1905, int(tot))
	t.AssertTrue(pieceCID.Equals(pieceCIDB))

	publicPieces := []abi.PieceInfo{{
		Size:     abi.UnpaddedPieceSize(127).Padded(),
		PieceCID: pieceCIDA,
	}, {
		Size:     abi.UnpaddedPieceSize(1016).Padded(),
		PieceCID: pieceCIDB,
	}}

	preGeneratedUnsealedCID, err := GenerateUnsealedCID(sealProofType, publicPieces)
	t.RequireNoError(err)

	// pre-commit the sector
	sealPreCommitPhase1Output, err := SealPreCommitPhase1(sealProofType, sectorCacheDirPath, stagedSectorFile.Name(), sealedSectorFile.Name(), sectorNum, minerID, ticket, publicPieces)
	t.RequireNoError(err)

	sealedCID, unsealedCID, err := SealPreCommitPhase2(sealPreCommitPhase1Output, sectorCacheDirPath, sealedSectorFile.Name())
	t.RequireNoError(err)

	t.AssertTrue(unsealedCID.Equals(preGeneratedUnsealedCID), "prover and verifier should agree on data commitment")

	// commit the sector
	sealCommitPhase1Output, err := SealCommitPhase1(sealProofType, sealedCID, unsealedCID, sectorCacheDirPath, sealedSectorFile.Name(), sectorNum, minerID, ticket, seed, publicPieces)
	t.RequireNoError(err)

	proof, err := SealCommitPhase2(sealCommitPhase1Output, sectorNum, minerID)
	t.RequireNoError(err)

	// verify the 'ole proofy
	isValid, err := VerifySeal(prf.SealVerifyInfo{
		SectorID: abi.SectorID{
			Miner:  minerID,
			Number: sectorNum,
		},
		SealedCID:             sealedCID,
		SealProof:             sealProofType,
		Proof:                 proof,
		DealIDs:               []abi.DealID{},
		Randomness:            ticket,
		InteractiveRandomness: seed,
		UnsealedCID:           unsealedCID,
	})
	t.RequireNoError(err)
	t.RequireTrue(isValid, "proof wasn't valid")

	// unseal the entire sector and verify that things went as we planned
	_, err = sealedSectorFile.Seek(0, 0)
	t.RequireNoError(err)
	t.RequireNoError(Unseal(sealProofType, sectorCacheDirPath, sealedSectorFile, unsealOutputFileA, sectorNum, minerID, ticket, unsealedCID))
	_, err = unsealOutputFileA.Seek(0, 0)
	t.RequireNoError(err)
	contents, err := ioutil.ReadFile(unsealOutputFileA.Name())
	t.RequireNoError(err)

	// unsealed sector includes a bunch of alignment NUL-bytes
	alignment := make([]byte, 889)

	// verify that we unsealed what we expected to unseal
	t.AssertTrue(bytes.Equal(someBytes[0:127], contents[0:127]), "bytes aren't equal")
	t.AssertTrue(bytes.Equal(alignment, contents[127:1016]), "bytes aren't equal")
	t.AssertTrue(bytes.Equal(someBytes[0:1016], contents[1016:2032]), "bytes aren't equal")

	// unseal just the first piece
	_, err = sealedSectorFile.Seek(0, 0)
	t.RequireNoError(err)
	err = UnsealRange(sealProofType, sectorCacheDirPath, sealedSectorFile, unsealOutputFileB, sectorNum, minerID, ticket, unsealedCID, 0, 127)
	t.RequireNoError(err)
	_, err = unsealOutputFileB.Seek(0, 0)
	t.RequireNoError(err)
	contentsB, err := ioutil.ReadFile(unsealOutputFileB.Name())
	t.RequireNoError(err)
	t.AssertEqual(127, len(contentsB))
	t.AssertTrue(bytes.Equal(someBytes[0:127], contentsB[0:127]), "bytes aren't equal")

	// unseal just the second piece
	_, err = sealedSectorFile.Seek(0, 0)
	t.RequireNoError(err)
	err = UnsealRange(sealProofType, sectorCacheDirPath, sealedSectorFile, unsealOutputFileC, sectorNum, minerID, ticket, unsealedCID, 1016, 1016)
	t.RequireNoError(err)
	_, err = unsealOutputFileC.Seek(0, 0)
	t.RequireNoError(err)
	contentsC, err := ioutil.ReadFile(unsealOutputFileC.Name())
	t.RequireNoError(err)
	t.AssertEqual(1016, len(contentsC))
	t.AssertTrue(bytes.Equal(someBytes[0:1016], contentsC[0:1016]), "bytes aren't equal")

	// verify that the sector builder owns no sealed sectors
	var sealedSectorPaths []string
	t.RequireNoError(filepath.Walk(sealedSectorsDir, visit(&sealedSectorPaths)))
	t.AssertEqual(1, len(sealedSectorPaths), sealedSectorPaths)

	// no sector cache dirs, either
	var sectorCacheDirPaths []string
	t.RequireNoError(filepath.Walk(sectorCacheRootDir, visit(&sectorCacheDirPaths)))
	t.AssertEqual(1, len(sectorCacheDirPaths), sectorCacheDirPaths)

	// run the FauxRep routine, for good measure
	fauxSectorCID, err := FauxRep(sealProofType, fauxSectorCacheDirPath, fauxSealedSectorFile.Name())
	t.RequireNoError(err, "FauxRep produced an error")
	t.RequireTrue(!cid.Undef.Equals(fauxSectorCID), "faux sector CID shouldn't be undefined")

	// run the FauxRep2 routine, for good measure
	fauxSectorCID2, err := FauxRep2(sealProofType, fauxSectorCacheDirPath, fauxSealedSectorFile.Name())
	t.RequireNoError(err, "FauxRep2 produced an error")
	t.RequireTrue(!cid.Undef.Equals(fauxSectorCID2), "faux sector CID 2 shouldn't be undefined")

	// generate a PoSt over the proving set before importing, just to exercise
	// the new API
	privateInfo := NewSortedPrivateSectorInfo(PrivateSectorInfo{
		SectorInfo: prf.SectorInfo{
			SectorNumber: sectorNum,
			SealedCID:    sealedCID,
		},
		CacheDirPath:     sectorCacheDirPath,
		PoStProofType:    winningPostProofType,
		SealedSectorPath: sealedSectorFile.Name(),
	})

	provingSet := []prf.SectorInfo{{
		SealProof:    sealProofType,
		SectorNumber: sectorNum,
		SealedCID:    sealedCID,
	}}

	// figure out which sectors have been challenged
	indicesInProvingSet, err := GenerateWinningPoStSectorChallenge(winningPostProofType, minerID, randomness[:], uint64(len(provingSet)))
	t.RequireNoError(err)

	var challengedSectors []prf.SectorInfo
	for idx := range indicesInProvingSet {
		challengedSectors = append(challengedSectors, provingSet[indicesInProvingSet[idx]])
	}

	proofs, err := GenerateWinningPoSt(minerID, privateInfo, randomness[:])
	t.RequireNoError(err)

	isValid, err = VerifyWinningPoSt(prf.WinningPoStVerifyInfo{
		Randomness:        randomness[:],
		Proofs:            proofs,
		ChallengedSectors: challengedSectors,
		Prover:            minerID,
	})
	t.RequireNoError(err)
	t.AssertTrue(isValid, "VerifyWinningPoSt rejected the (standalone) proof as invalid")
}

func WorkflowGetGPUDevicesDoesNotProduceAnError(t TestHelper) {
	devices, err := GetGPUDevices()
	t.RequireNoError(err)
	fmt.Printf("devices: %+v\n", devices) // clutters up test output, but useful
}

func WorkflowRegisteredSealProofFunctions(t TestHelper) {
	sealTypes := []abi.RegisteredSealProof{
		abi.RegisteredSealProof_StackedDrg2KiBV1,
		abi.RegisteredSealProof_StackedDrg8MiBV1,
		abi.RegisteredSealProof_StackedDrg512MiBV1,
		abi.RegisteredSealProof_StackedDrg32GiBV1,
		abi.RegisteredSealProof_StackedDrg64GiBV1,

		abi.RegisteredSealProof_StackedDrg2KiBV1_1,
		abi.RegisteredSealProof_StackedDrg8MiBV1_1,
		abi.RegisteredSealProof_StackedDrg512MiBV1_1,
		abi.RegisteredSealProof_StackedDrg32GiBV1_1,
		abi.RegisteredSealProof_StackedDrg64GiBV1_1,
	}

	for _, st := range sealTypes {
		v, err := GetSealVersion(st)
		t.AssertNoError(err)
		t.AssertTrue(len(v) > 0)
	}
}

func WorkflowRegisteredPoStProofFunctions(t TestHelper) {
	postTypes := []abi.RegisteredPoStProof{
		abi.RegisteredPoStProof_StackedDrgWinning2KiBV1,
		abi.RegisteredPoStProof_StackedDrgWinning8MiBV1,
		abi.RegisteredPoStProof_StackedDrgWinning512MiBV1,
		abi.RegisteredPoStProof_StackedDrgWinning32GiBV1,
		abi.RegisteredPoStProof_StackedDrgWinning64GiBV1,

		abi.RegisteredPoStProof_StackedDrgWindow2KiBV1,
		abi.RegisteredPoStProof_StackedDrgWindow8MiBV1,
		abi.RegisteredPoStProof_StackedDrgWindow512MiBV1,
		abi.RegisteredPoStProof_StackedDrgWindow32GiBV1,
		abi.RegisteredPoStProof_StackedDrgWindow64GiBV1,
	}

	for _, pt := range postTypes {
		v, err := GetPoStVersion(pt)
		t.AssertNoError(err)
		t.AssertTrue(len(v) > 0)
	}
}

func WorkflowGenerateWinningPoStSectorChallengeEdgeCase(t TestHelper) {
	for i := 0; i < 10000; i++ {
		var randomnessFr32 [32]byte
		_, err := io.ReadFull(rand.Reader, randomnessFr32[0:31]) // last byte of the 32 is always NUL
		t.RequireNoError(err)

		minerID := randActorID()
		eligibleSectorsLen := uint64(1)

		indices2, err := GenerateWinningPoStSectorChallenge(abi.RegisteredPoStProof_StackedDrgWinning2KiBV1, minerID, randomnessFr32[:], eligibleSectorsLen)
		t.RequireNoError(err)
		t.AssertEqual(1, len(indices2))
		t.AssertEqual(0, int(indices2[0]))
	}
}

func WorkflowGenerateWinningPoStSectorChallenge(t TestHelper) {
	for i := 0; i < 10000; i++ {
		var randomnessFr32 [32]byte
		_, err := io.ReadFull(rand.Reader, randomnessFr32[0:31]) // last byte of the 32 is always NUL
		t.RequireNoError(err)

		minerID := randActorID()
		eligibleSectorsLen := randUInt64()

		if eligibleSectorsLen == 0 {
			continue // no fun
		}

		indices, err := GenerateWinningPoStSectorChallenge(abi.RegisteredPoStProof_StackedDrgWinning2KiBV1, minerID, randomnessFr32[:], eligibleSectorsLen)
		t.AssertNoError(err)

		max := uint64(0)
		for idx := range indices {
			if indices[idx] > max {
				max = indices[idx]
			}
		}

		t.AssertTrue(max < eligibleSectorsLen, "out of range value - max: ", max, "eligibleSectorsLen: ", eligibleSectorsLen)
		t.AssertTrue(uint64(len(indices)) <= eligibleSectorsLen, "should never generate more indices than number of eligible sectors")
	}
}

func randActorID() abi.ActorID {
	bID, err := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
	if err != nil {
		panic(err)
	}
	return abi.ActorID(bID.Uint64())
}

func randUInt64() uint64 {
	buf := make([]byte, 8)
	_, err := rand.Read(buf)
	if err != nil {
		panic(err)
	}

	return binary.LittleEndian.Uint64(buf)
}

func requireTempFile(t TestHelper, fileContentsReader io.Reader, size uint64) *os.File {
	file, err := ioutil.TempFile("", "")
	t.RequireNoError(err)

	written, err := io.Copy(file, fileContentsReader)
	t.RequireNoError(err)
	// check that we wrote everything
	t.RequireEqual(int(size), int(written))

	t.RequireNoError(file.Sync())

	// seek to the beginning
	_, err = file.Seek(0, 0)
	t.RequireNoError(err)

	return file
}

func requireTempDirPath(t TestHelper, prefix string) string {
	dir, err := ioutil.TempDir("", prefix)
	t.RequireNoError(err)

	return dir
}

func visit(paths *[]string) filepath.WalkFunc {
	return func(path string, info os.FileInfo, err error) error {
		if err != nil {
			panic(err)
		}
		*paths = append(*paths, path)
		return nil
	}
}

type TestHelper interface {
	AssertEqual(expected, actual interface{}, msgAndArgs ...interface{}) bool
	AssertNoError(err error, msgAndArgs ...interface{}) bool
	AssertTrue(value bool, msgAndArgs ...interface{}) bool
	RequireEqual(expected interface{}, actual interface{}, msgAndArgs ...interface{})
	RequireNoError(err error, msgAndArgs ...interface{})
	RequireTrue(value bool, msgAndArgs ...interface{})
}
